src/ports/postgres/modules/graph/pagerank.py_in

```diff
@@ -158,44 +313,198 @@ def pagerank(schema_madlib, vertex_table, vertex_id, edge_table, edge_args,
         # https://en.wikipedia.org/wiki/PageRank#Damping_factor
 
         # The query below computes the PageRank of each node using the above formula.
+        # A small explanatory note on ignore_group_clause:
+        # This is used only when grouping is set. This essentially will have
+        # the condition that will help skip the PageRank computation on groups
+        # that have converged.
         plpy.execute("""
                 CREATE TABLE {message} AS
-                SELECT {edge_temp_table}.{dest} AS {vertex_id},
-                        SUM({v1}.pagerank/{out_cnts}.{out_cnts_cnt})*{damping_factor}+{random_prob} AS pagerank
+                SELECT {grouping_cols_select} {edge_temp_table}.{dest} AS {vertex_id},
+                        SUM({v1}.pagerank/{out_cnts}.{out_cnts_cnt})*{damping_factor}+{random_jump_prob} AS pagerank
                 FROM {edge_temp_table}
-                    INNER JOIN {cur} ON {edge_temp_table}.{dest}={cur}.{vertex_id}
-                    INNER JOIN {out_cnts} ON {out_cnts}.{vertex_id}={edge_temp_table}.{src}
-                    INNER JOIN {cur} AS {v1} ON {v1}.{vertex_id}={edge_temp_table}.{src}
-                GROUP BY {edge_temp_table}.{dest}
-            """.format(**locals()))
+                    INNER JOIN {cur} ON {cur_join_clause}
+                    INNER JOIN {out_cnts} ON {out_cnts_join_clause}
+                    INNER JOIN {cur} AS {v1} ON {v1_join_clause}
+                    {vertices_per_group_inner_join}
+                {ignore_group_clause}
+                GROUP BY {grouping_cols_select} {edge_temp_table}.{dest}
+            """.format(grouping_cols_select=edge_grouping_cols_select+', '
+                    if grouping_cols else '',
+                random_jump_prob='MIN({vpg}.{random_prob})'.format(**locals())
+                    if grouping_cols else random_probability,
+                vertices_per_group_inner_join="""INNER JOIN {vertices_per_group}
+                    AS {vpg} ON {vpg_join_clause}""".format(**locals())
+                    if grouping_cols else '',
+                ignore_group_clause=' WHERE '+get_ignore_groups(
+                    summary_table, edge_temp_table, grouping_cols_list)
+                    if iteration_num>0 and grouping_cols else '',
+                **locals()))
         # If there are nodes that have no incoming edges, they are not captured in the message table.
         # Insert entries for such nodes, with random_prob.
         plpy.execute("""
                 INSERT INTO {message}
-                SELECT {vertex_id}, {random_prob}::DOUBLE PRECISION AS pagerank
-                FROM {cur}
-                WHERE {vertex_id} NOT IN (
+                SELECT {grouping_cols_select} {cur}.{vertex_id}, {random_jump_prob} AS pagerank
+                FROM {cur} {vpg_from_clause}
+                WHERE {vpg_where_clause} {vertex_id} NOT IN (
```

Comment:
The query does takes care of the scenario you mentioned. The `{message_grp_where}` clause inside NOT IN checks if `cur.grouping_cols=message.grouping_cols`. I also tried an example to confirm it does work as expected. So leaving this as is.